flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -74,88 +76,233 @@
+			} else if (currentPattern.getQuantifier() == Quantifier.TIMES) {
+				if (currentPattern.getTimes() > 1) {
+					final State<T> timesState = new State<>(currentPattern.getName(), State.StateType.Normal);
+					states.add(timesState);
+					sinkState = convertToTimesState(sinkState, timesState, currentPattern.getTimes() - 1);
 				}
+				beginningState = new State<>(currentPattern.getName(), State.StateType.Start);
+				beginningState.addTake(sinkState, (FilterFunction<T>) currentPattern.getFilterFunction());
+			} else {
+				beginningState = new State<>(currentPattern.getName(), State.StateType.Start);
+				beginningState.addTake(sinkState, (FilterFunction<T>) currentPattern.getFilterFunction());
 			}
 
-			// add the beginning state
-			final State<T> beginningState;
+			states.add(beginningState);
+			usedNames.add(beginningState.getName());
+
+			return beginningState;
+		}
+
+		private State<T> convertToTimesState(State<T> sinkState, State<T> sourceState, int times) {
+			convertToSingletonState(sinkState, sourceState);
+			for (int i = 0; i < times - 1; i++) {
+				sinkState = sourceState;
+				sourceState = new State<>(currentPattern.getName(), State.StateType.Normal);
+				states.add(sourceState);
+				convertToSingletonState(sinkState, sourceState);
+			}
+			return sourceState;
+		}
+
+		@SuppressWarnings("unchecked")
+		private void convertToSingletonState(
+			final State<T> sinkState,
+			final State<T> sourceState) {
+
+			final FilterFunction<T> currentFilterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
+			final FilterFunction<T> trueFunction = FilterFunctions.trueFunction();
+			sourceState.addTake(sinkState, currentFilterFunction);
+
+			final State<T> ignoreState;
+			if (currentPattern.getQuantifier() == Quantifier.OPTIONAL) {
+				sourceState.addProceed(sinkState, trueFunction);
+				ignoreState = new State<>(currentPattern.getName(), State.StateType.Normal);
 
-			if (states.containsKey(BEGINNING_STATE_NAME)) {
-				beginningState = states.get(BEGINNING_STATE_NAME);
+				ignoreState.addTake(sinkState, currentFilterFunction);
+				states.add(ignoreState);
 			} else {
-				beginningState = new State<>(BEGINNING_STATE_NAME, State.StateType.Start);
-				states.put(BEGINNING_STATE_NAME, beginningState);
+				ignoreState = sourceState;
+			}
+
+			if (currentPattern instanceof FollowedByPattern) {
+				sourceState.addIgnore(ignoreState, trueFunction);
+			}
+		}
+
+		@SuppressWarnings("unchecked")
+		private State<T> createFirstMandatoryStateOfLoop(
+			final State<T> sinkState,
+			final State.StateType stateType) {
+
+			final FilterFunction<T> currentFilterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
+			final State<T> firstState = new State<>(currentPattern.getName(), stateType);
+
+			firstState.addTake(sinkState, currentFilterFunction);
+			if (currentPattern instanceof FollowedByPattern) {
+				if (currentPattern.getQuantifier() == Quantifier.ONE_OR_MORE_COMBINATIONS) {
```

Comment:
In the `Quantifier` you can add a flag like `isCombinations` for the `ONE_OR_MORE_COMBINATIONS` and the `ZERO_OR_MORE_COMBINATIONS` and check it here. Given that you already check for `atLeastOne` in the check I proposed above, this will be sufficient.